GVFS/GVFS.UnitTests.Windows/Windows/Upgrader/UpgradeOrchestratorTests.cs

```diff
@@ -0,0 +1,243 @@
+﻿using GVFS.Common;
+using GVFS.Common.Git;
+using GVFS.Tests.Should;
+using GVFS.UnitTests.Mock.Common;
+using GVFS.UnitTests.Windows.Mock.Upgrader;
+using GVFS.Upgrader;
+using Moq;
+using NUnit.Framework;
+using System;
+using System.Collections.Generic;
+
+namespace GVFS.UnitTests.Upgrader
+{
+    [TestFixture]
+    public class UpgradeOrchestratorTests
+    {
+        private const string GVFSVersion = "1.1.18115.1";
+
+        private delegate void TryGetNewerVersionCallback(out System.Version version, out string message);
+        private delegate void UpgradeAllowedCallback(out string message);
+        private delegate void TryRunPreUpgradeChecksCallback(out string delegateMessage);
+        private delegate void TryDownloadNewestVersionCallback(out string message);
+        private delegate void TryRunInstallerCallback(InstallActionWrapper installActionWrapper, out string error);
+
+        private MockTracer Tracer { get; set; }
+        private MockTextWriter Output { get; set; }
+        private MockInstallerPrerunChecker PreRunChecker { get; set; }
+        private Mock<LocalGVFSConfig> MoqLocalConfig { get; set; }
+        private Mock<IProductUpgrader> MoqUpgrader { get; set; }
+
+        private UpgradeOrchestrator orchestrator { get; set; }
+
+        [SetUp]
+        public void Setup()
+        {
+            this.Tracer = new MockTracer();
+            this.Output = new MockTextWriter();
+            this.PreRunChecker = new MockInstallerPrerunChecker(this.Tracer);
+            this.PreRunChecker.Reset();
+            this.MoqUpgrader = this.DefaultUpgrader();
+            this.orchestrator = new UpgradeOrchestrator(
+                this.MoqUpgrader.Object,
+                this.Tracer,
+                this.PreRunChecker,
+                input: null,
+                output: this.Output);
+
+            this.SetUpgradeAvailable(new Version(GVFSVersion), error: null);
+        }
+
+        [TestCase]
+        public void ExecuteSucceedsWhenUpgradeAvailable()
+        {
+            this.orchestrator.Execute();
+
+            this.VerifyOrchestratorInvokes(
+                upgradeAllowed: true,
+                queryNewestVersion: true,
+                downloadNewestVersion: true,
+                installNewestVersion: true,
+                cleanup: true);
+
+            this.orchestrator.ExitCode.ShouldEqual(ReturnCode.Success);
+        }
+
+        [TestCase]
+        public void NewVersionNotAvailable()
```

Comment:
Two minor naming suggestions:

- Like the above test case, start the method name with `ExecuteSucceedsWhen`
- Does `NewVersionNotAvailable` mean there is a new version but VFS4G fails to download it?  Or does it mean there is no new version on the server?  If it's the second, consider changing the name to something like `ExecuteSucceedsWhenOnLatestVersion`